src/handlers.rs

```diff
@@ -0,0 +1,316 @@
+    name = "Topic handler",
+    skip(bot, msg, db, scheduler),
+)]
+async fn topic_handler(
+    bot: Bot,
+    msg: Message,
+    mut db: Database,
+    scheduler: Scheduler,
+) -> HandlerResult {
+    let thread_id = msg.thread_id.expect("infallible").0.0;
+    let mut mapping = db.get_mapping(thread_id as i64).await?.ok_or_else(|| {
+        tracing::warn!("Mapping not configured: {thread_id}");
+        "Mapping not configured"
+    })?;
+    let with_reply = msg.reply_to_message()
+        .map_or(false, |reply| reply.id.0 != thread_id);
+    
+    let last_private = if with_reply {
+        let reply_to_message_id = msg.reply_to_message().expect("infallible").id.0;
+        let shift = msg.id.0 - reply_to_message_id - 1;
+        let reply_msg_id = MessageId(mapping.last_private.0 - shift);
+        bot.copy_message(mapping.recipient_chat, msg.chat.id, msg.id)
+            .reply_parameters(ReplyParameters::new(reply_msg_id))
+            .await?
+    } else {
+        bot.copy_message(mapping.recipient_chat, msg.chat.id, msg.id).await?
+    };
+    mapping.sync(last_private, msg.id);
+    db.sync_mapping(mapping, scheduler).await?;
+
+    Ok(())
+}
+
+#[instrument(
+    name = "Admin command handler",
+    skip(bot, msg, cmd, forum_id, db, scheduler),
+)]
+async fn admin_command_handler(
+    bot: Bot, 
+    msg: Message,
+    cmd: AdminCommand,  // while 1 command !!!
+    forum_id: ChatId, 
+    mut db: Database,
+    scheduler: Scheduler,
+) -> HandlerResult {
+    if let AdminCommand::DropTopic(forum_name) = cmd {
+        let thread_id = msg.thread_id.expect("infallible");
+        if forum_name.is_empty() {
+            bot.send_message(
+                msg.chat.id, 
+                "⚠️ Please, specify a new topic name,\nf.e. /drop_topic {topic_name}"
+            )
+                .message_thread_id(thread_id).await?;
+            return Ok(());
+        }
+        let thread_id_num = thread_id.0.0 as i64;
+        if let Some(mapping) = db.get_mapping(thread_id_num).await? {
+            // Delete mapping
+            let _ = db.drop_mapping(thread_id_num).await;
+            scheduler.cancel_task(mapping.unique_id() as u64); // Cancel scheduled synchronization
+            // Drop topic
+            let forum_name = format!("🗄 {forum_name}");
+            close_topic(&bot, forum_id, thread_id, &forum_name).await?;
+            bot.send_message(msg.chat.id, "🗑 Topic dropped")
+                .message_thread_id(thread_id).await?;
+            tracing::info!("Topic dropped: {}", thread_id.0.0);
+        }
+    }
+    
+    Ok(())
+}
+
+#[instrument(
+    name = "Ban handler",
+    skip(bot, call, db, forum_id, scheduler),
+)]
+async fn ban_handler(
+    bot: Bot, 
+    call: CallbackQuery,
+    mut db: Database, 
+    forum_id: ChatId,
+    scheduler: Scheduler,
+) -> HandlerResult {
+    let init_msg_binding = call.message.expect("infallible");
+    let init_msg = init_msg_binding.regular_message().expect("infallible");
```

Comment:
`"infallible"` is not informative. In this case, functions alike rely on invariants enforced somewhere else. Where, exactly? Well, at the call site.

This is by no means a fatal flaw, since in the end - you're unlikely to reuse these handlers, or miss invariant enforcement; but it would be much nicer if the **type system itself** was proving all of that.

What you _could_ do is to ask for things you _actually need_ in each handler (like, whatever the type of `init_msg` is here); that would make handler itself truly infallible, as there would be nothing to fail.

As a nice side-effect - that would make a functions easier to test, as they would require simpler types.

Reply:
If you're talking about adding dependencies to arguments, I've tried it, but failed miserably. As I understand it, teloxide is not smart enough to search the Update internals for the data I want to have ready to use. Although there is an idea to do all the dirty work with unpacking in filters and call the handler from there, but it won't solve the problem. But to write different reasons in .except methods is a good idea, I'll do it